src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs

```diff
@@ -731,9 +732,12 @@ public override ValueGetter<TValue> GetGetter<TValue>(DataViewSchema.Column colu
             {
                 Ch.CheckParam(column.Index < _colToActivesIndex.Length, nameof(column));
                 Ch.CheckParam(_colToActivesIndex[column.Index] >= 0, nameof(column), "requested column not active");
-                ValueGetter<TValue> getter = _getters[_colToActivesIndex[column.Index]] as ValueGetter<TValue>;
+
+                var originGetter = _getters[_colToActivesIndex[column.Index]];
+                ValueGetter<TValue> getter = originGetter as ValueGetter<TValue>;
                 if (getter == null)
-                    throw Ch.Except("Invalid TValue: '{0}'", typeof(TValue));
+                    throw Ch.Except($"Invalid TValue: '{typeof(TValue)}', " +
+                            $"expected type: '{originGetter.GetType().GetGenericArguments().First()}'.");
```

Comment:
_(pinned to code to allow for threaded conversation)_

In general, I think that the approach you used to fix this issue is fine, and it's actually a very good addition to make our exception messages more clear.

The only potential problem I see with it is that _maybe_ this will end up throwing _another_ exception: for example, the `_getters` in this method is an array of `Delegates` (i.e. it's not an array of `ValueGetter<>`'s) so the cast `as ValueGetter<TValue>` could potentially fail not only if the the generic parameter is mistaken, but also in the case that `_getters[i]` isn't a `ValueGetter<>` to begin with. If it isn't a `ValueGetter<>`, and if it's not a delegate with generic types, then the `.First()` call in here will throw an exception because the `.GetGenericArguments()` method would have thrown an empty array.

I haven't checked the other places were you've made changes, but some of them do have an original `_getter` which is already a `ValueGetter<>` (so this wouldn't be a problem in those cases) but in other cases the original `_getter` is a `Delegate` instead of `ValueGetter<>` (so potentially they might not have GenericArguments, and then throw another exception).

I'm not sure if I'm explaining myself.

In any case, I also don't know if we should worry about this problem... that other exception would only be thrown in some fringe case (which I'm not sure if it's actually possible to reach that case), and in any way we were already throwing an exception. So maybe this is unimportant.

Reply:
I understand your concern, please check out the implementation of this _getters, mainly method CreateGetterDelegate in same class and you will find below implementation, so it is guaranteed you will get ValueGetter<TValue> for this Delegate but the real problem is you are using wrong TValue type.

            private ValueGetter<TValue> CreateGetterDelegate<TValue>(ShufflePipe pipe)
            {
                Ch.AssertValue(pipe);
                Ch.Assert(pipe is ShufflePipe<TValue>);
                var pipe2 = (ShufflePipe<TValue>)pipe;
                ValueGetter<TValue> getter =
                    (ref TValue value) =>
                    {
                        Ch.Assert(_pipeIndex == _pipeIndices[_circularIndex]);
                        pipe2.Fetch(_pipeIndex, ref value);
                    };
                return getter;
            }


---
In reply to: [433925403](https://github.com/dotnet/machinelearning/pull/5189#discussion_r433925403) [](ancestors = 433925403)